mgit/lib/index.js

```diff
@@ -0,0 +1,110 @@
+/**
+ * @license Copyright (c) 2003-2017, CKSource - Frederico Knabben. All rights reserved.
+ * For licensing, see LICENSE.md.
+ */
+
+'use strict';
+
+const fs = require( 'fs' );
+const path = require( 'path' );
+const ProgressBar = require( 'progress' );
+const createForkPool = require( './createforkpool' );
+
+module.exports = function( parameters, options ) {
+	// options.repositoryResolver should be an absolute path.
+	if ( fs.existsSync( options.repositoryResolver ) ) {
+		options.repositoryResolver = path.resolve( options.repositoryResolver );
+	}
+
+	const mgitConf = require( path.resolve( 'mgit.json' ) );
+	const forkPool = createForkPool( path.join( __dirname, 'forkexecutable.js' ) );
+
+	const commandPath = path.join( __dirname, 'commands', parameters[ 0 ] );
+	const command = require( commandPath );
+
+	if ( typeof command.beforeExecute === 'function' ) {
+		command.beforeExecute();
+	}
+
+	const parsedPackages = new Set();
+	const errors = new Set();
+
+	console.time( 'Processing time' );
+
+	const packageNames = Object.keys( mgitConf.dependencies );
+
+	const progress = new ProgressBar( '[:bar] :current/:total (:percent)', {
+		total: packageNames.length,
+		complete: '=',
+		incomplete: ' ',
+		width: 75
+	} );
+
+	for ( const item of packageNames ) {
+		enqueue( item );
+	}
+
+	function enqueue( packageName ) {
+		if ( parsedPackages.has( packageName ) ) {
+			progress.tick();
+
+			return;
+		}
+
+		parsedPackages.add( packageName );
+
+		const data = {
+			command: commandPath,
+			parameters: parameters.slice( 1 ),
+			options,
+			name: packageName,
+			mgit: mgitConf
+		};
+
+		forkPool.enqueue( data )
+			.then( ( response ) => {
+				progress.tick();
+
+				if ( Array.isArray( response ) ) {
+					progress.total += response.length;
```

Comment:
Please structure the `response` a bit more. The bootstrap command should return deps of a handled package as `response.dependencies` or something. Otherwise, this is nightmare to understand :P